src/main/java/com/hubspot/jinjava/lib/tag/SetTag.java

```diff
@@ -81,23 +84,28 @@ public String interpret(TagNode tagNode, JinjavaInterpreter interpreter) {
 
     String[] varTokens = var.split(",");
 
-    if (varTokens.length > 1) {
-      // handle multi-variable assignment
-      @SuppressWarnings("unchecked")
-      List<Object> exprVals = (List<Object>) interpreter.resolveELExpression("[" + expr + "]", tagNode.getLineNumber());
+    try {
+      if (varTokens.length > 1) {
+        // handle multi-variable assignment
+        @SuppressWarnings("unchecked")
+        List<Object> exprVals = (List<Object>) interpreter.resolveELExpression("[" + expr + "]", tagNode.getLineNumber());
 
-      if (varTokens.length != exprVals.size()) {
-        throw new TemplateSyntaxException(tagNode.getMaster().getImage(), "Tag 'set' declares an uneven number of variables and assigned values", tagNode.getLineNumber(), tagNode.getStartPosition());
-      }
+        if (varTokens.length != exprVals.size()) {
+          throw new TemplateSyntaxException(tagNode.getMaster().getImage(), "Tag 'set' declares an uneven number of variables and assigned values", tagNode.getLineNumber(), tagNode.getStartPosition());
+        }
 
-      for (int i = 0; i < varTokens.length; i++) {
-        String varItem = varTokens[i].trim();
-        interpreter.getContext().put(varItem, exprVals.get(i));
-      }
+        for (int i = 0; i < varTokens.length; i++) {
+          String varItem = varTokens[i].trim();
+          interpreter.getContext().put(varItem, exprVals.get(i));
+        }
 
-    } else {
-      // handle single variable assignment
-      interpreter.getContext().put(var, interpreter.resolveELExpression(expr, tagNode.getLineNumber()));
+      } else {
+        // handle single variable assignment
+        interpreter.getContext().put(var, interpreter.resolveELExpression(expr, tagNode.getLineNumber()));
+      }
+    } catch (DeferredValueException e) {
+      Stream.of(varTokens).forEach(varToken -> interpreter.getContext().put(varToken.trim(), DeferredValue.instance()));
```

Comment:
This is the key bit that makes "deferred-ness" infectious, for any other readers trying to see what actually changed through the whitespace diff.